### PR TITLE
Ft/s3 c 1155 put bucket lifecycle

### PR DESCRIPTION
**Pull request by dora-korpar**
Implements the S3 API for put bucket lifecycle

Depends on https://github.com/scality/Arsenal/pull/389

### CODE REVIEW

File: lib/api/bucketPutLifecycle.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    require('arsenal').models.LifecycleConfiguration;
+
+const parseXML = require('../utilities/parseXML');
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+const metadata = require('../metadata/wrapper');
+const { metadataValidateBucket } = require('../metadata/metadataUtils');
+const { pushMetric } = require('../utapi/utilities');
+
+/**
+ * Bucket Put Versioning - Create or update bucket lifecycle configuration
```
Comment (bennettbuchanan) [inline lib/api/bucketPutLifecycle.js]: ~Bucket Put Versioning~ bucketPutLifecycle

File: lib/api/bucketPutLifecycle.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        bucketName,
+        requestType: 'bucketOwnerAction',
+    };
+    return waterfall([
+        next => parseXML(request.post, log, next),
+        (parsedXml, next) => {
+            const lcConfigClass = new LifecycleConfiguration(parsedXml);
+            // if there was an error getting lifecycle configuration,
+            // returned configObj will contain 'error' key
+            process.nextTick(() => {
```
Comment (bennettbuchanan) [inline lib/api/bucketPutLifecycle.js]: Just curious why this needs to happen at next cycle?

File: lib/api/bucketPutLifecycle.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                const configObj = lcConfigClass.getLifecycleConfiguration();
+                return next(configObj.error || null, configObj);
+            });
+        },
+        (lcConfig, next) => metadataValidateBucket(metadataValParams, log,
+            (err, bucket) => {
+                if (err) {
+                    return next(err, bucket);
```
Comment (bennettbuchanan) [inline lib/api/bucketPutLifecycle.js]: No need for second arg to `next` here for error condition.

File: tests/functional/aws-node-sdk/test/bucket/putBucketLifecycle.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                assertError(err, null, done));
+        });
+
+        it('should not allow lifecycle config with no Status', done => {
+            const params = getLifecycleParams({ key: 'Status', value: '' });
+            s3.putBucketLifecycleConfiguration(params, err =>
+                assertError(err, 'MalformedXML', done));
+        });
+
+        it('should not allow lifecycle conifg with no Prefix or Filter',
```
Comment (bennettbuchanan) [inline tests/functional/aws-node-sdk/test/bucket/putBucketLifecycle.js]: ~conifg~ config

File: lib/api/bucketPutLifecycle.js (hunk 3 above)
Comment (bennettbuchanan) [reply lib/api/bucketPutLifecycle.js]: Oh wait, I guess this is for collecting CORS headers at the bottom?

Comment (dora-korpar) [reply lib/api/bucketPutLifecycle.js]: yup, for the corsHeaders

File: lib/api/bucketPutLifecycle.js (hunk 2 above)
Comment (dora-korpar) [reply lib/api/bucketPutLifecycle.js]: getting the config is a synchronous call, so I believe it's just best practice to use `process.nextTick()` to mimic asynchronicity

**Comment by dora-korpar** [comment]
@ironman-machine try

**Comment by dora-korpar** [comment]
@ironman-machine try

**Comment by dora-korpar** [comment]
@ironman-machine try

File: tests/functional/aws-node-sdk/test/bucket/putBucketLifecycle.js [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+// Check for the expected error response code and status code.
+function assertError(err, expectedErr, cb) {
+    if (expectedErr === null) {
+        assert.strictEqual(err, null, `expected no error but got '${err}'`);
+    } else {
+        assert.strictEqual(err.code, expectedErr, 'incorrect error response ' +
+            `code: should be '${expectedErr}' but got '${err.code}'`);
+        assert.strictEqual(err.statusCode, errors[expectedErr].code,
+            'incorrect error status code: should be 400 but got ' +
```
Comment (jonathan-gramain) [inline tests/functional/aws-node-sdk/test/bucket/putBucketLifecycle.js]: instead of 400, you want to print `${errors[expectedErr].code}`?

**Comment by bennettbuchanan** [comment]
@ironman-machine r+

**Comment by bennettbuchanan** [comment]
The "r+" command was successful but merging failed due to pending ci/circleci_enterprise tests. Those tests have now completed successfully. Merging manually.

**Pull request merged**